pkg/connection/manager.go

```diff
@@ -0,0 +1,158 @@
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	"github.com/crossplane/crossplane-runtime/pkg/resource/unstructured"
+)
+
+// Error strings.
+const (
+	errConnectStore    = "cannot connect to secret store"
+	errWriteStore      = "cannot write to secret store"
+	errDeleteFromStore = "cannot delete from secret store"
+	errGetStoreConfig  = "cannot get store config"
+)
+
+// A StoreConfigKind contains the type metadata for a kind of StoreConfig
+// resource.
+type StoreConfigKind schema.GroupVersionKind
+
+// StoreBuilderFn is a function that builds and returns a Store with a given
+// store config.
+type StoreBuilderFn func(ctx context.Context, local client.Client, cfg v1.SecretStoreConfig) (Store, error)
+
+// A ManagerOption configures a Manager.
+type ManagerOption func(*Manager)
+
+// WithLogger specifies how the Manager should log messages.
+func WithLogger(l logging.Logger) ManagerOption {
+	return func(m *Manager) {
+		m.log = l
+	}
+}
+
+// WithStoreBuilder configures the StoreBuilder to use.
+func WithStoreBuilder(sb StoreBuilderFn) ManagerOption {
+	return func(m *Manager) {
+		m.storeBuilder = sb
+	}
+}
+
+// Manager is a connection details manager that satisfies the required
+// interfaces to work with connection details by managing interaction with
+// different store implementations.
+type Manager struct {
+	client         client.Client
+	newStoreConfig func() StoreConfig
+	storeBuilder   StoreBuilderFn
+
+	log logging.Logger
+}
+
+// NewManager returns a new connection Manager.
+func NewManager(c client.Client, of StoreConfigKind, o ...ManagerOption) *Manager {
+	nsc := func() StoreConfig {
+		return store.NewConfig(store.ConfigWithGroupVersionKind(schema.GroupVersionKind(of)))
+	}
+
+	m := &Manager{
+		client:         c,
+		newStoreConfig: nsc,
+		storeBuilder:   RuntimeStoreBuilder,
+
+		log: logging.NewNopLogger(),
+	}
+
+	for _, mo := range o {
+		mo(m)
+	}
+
+	return m
+}
+
+// PublishConnection publishes the supplied ConnectionDetails to a secret on
+// the configured connection Store.
+func (m *Manager) PublishConnection(ctx context.Context, mg resource.Managed, c managed.ConnectionDetails) error {
+	return m.publishConnection(ctx, mg.(SecretOwner), store.KeyValues(c))
+}
+
+// UnpublishConnection deletes connection details secret from the configured
+// connection Store.
+func (m *Manager) UnpublishConnection(ctx context.Context, mg resource.Managed, c managed.ConnectionDetails) error {
+	return m.unpublishConnection(ctx, mg.(SecretOwner), store.KeyValues(c))
+}
+
+func (m *Manager) connectStore(ctx context.Context, p *v1.PublishConnectionDetailsTo) (Store, error) {
+	sc := m.newStoreConfig()
+	if err := unstructured.NewClient(m.client).
+		Get(ctx, types.NamespacedName{Name: p.SecretStoreConfigRef.Name}, sc); err != nil {
+		return nil, errors.Wrap(err, errGetStoreConfig)
```

Comment:
Nit: I'd prefer to avoid wrapping the `Get` call here since it moves it to the same indentation level as the function logic. I don't think the line was too long anyhow.

Reply:
Actually, are you sure we need the `unstructured.NewClient` here in the first place? Could we not use the same pattern we use with `ProviderConfig` (e.g. like the `ProviderConfig` reconciler in this module) where the `StoreConfig` type is defined by each provider, so we can pass in a 'real' struct instead of a wrapper on `*unstructured.Unstructured`?

Reply:
Good point, followed the same approach as `ProviderConfig`.